### DISCUSSION TITLE
Solve large systems of equation with matrix regularization

### DISCUSSION THREAD

**Discussion by muhlbach** [cvxpy/cvxpy#1694] [Q&A]
Hi cvxpy maintainers and contributors! As always, I want to start out by highlighting how awesome work you're doing with this module--love it!

For the question. Assume that I have access to:
- (N x M)-dim matrix `Y`
- (N x d)-dim matrix `E`
- (M x d)-dim matrix `A`

Now, consider a (d x d)-dim decision matrix `R`. My _initial_ objective is to minimize the squared Frobenius norm:
 `R = argmin ||Y - E*R*A'||_F^2`. 

However, it turns out that without any further structure `R` estimated this way greatly overfits, and using unseen data it is much better to just an identity matrix as `R` (because `E` and `A` represent embeddings, and the cosine similarity should resemble `Y`. But the embeddings are unsupervised in the sense that they are raw embeddings some BERT-like model, and now I want to finetune (rotate) these embeddings to the space of `Y `.)

My first attempt to regularize `R` was to enforce it to be diagonal, essentially reducing the variables from d^2 to d. This is easily done in cvxpy (thank you once again!). However, it turns out that this regularization is not enough. It is still better to use the identity matrix as `R`.

My second attempt was to add an explicit penalty term, similar to:
 `R = argmin ||Y - E*R*A'||_F^2 + lambda*||R||_F^2`, where lambda is some scalar penalty.

Again, this is easily implemented in cvxpy, but it takes forever to run and I end up canceling the program before completion.

**Question: Any inputs on how to solve this optimization problem?**

See below for current code with data. It should run off-the-shelf but takes 5-7 minutes to complete.

```
import numpy as np
import cvxpy as cp
import scipy as sp
import pandas as pd

class FineTuner(object):
    """
    This class fine-tune embeddings
    """
    # -------------------------------------------------------------------------
    # Constructor function
    # -------------------------------------------------------------------------
    def __init__(self,
                 verbose=False,
                 nonneg=False,
                 nonpos=False,
                 symmetric=False,
                 diag=True,
                 PSD=False,
                 NSD=False,
                 boolean=False,
                 integer=False,
                 pos=False,
                 neg=False):        
        self.verbose = verbose
        self.nonneg = nonneg
        self.nonpos = nonpos
        self.symmetric = symmetric
        self.diag = diag
        self.PSD = PSD
        self.NSD = NSD
        self.boolean = boolean
        self.integer = integer
        self.pos = pos
        self.neg = neg
        
        # Instantiate
        self.problem_is_set_up = False

    # -------------------------------------------------------------------------
    # Class variables
    # -------------------------------------------------------------------------
    NORM = "fro"
    
    # -------------------------------------------------------------------------
    # Public
    # -------------------------------------------------------------------------
    def setup(self,Y,A):                

        # Get dimensions
        self.N = Y.shape[0]
        self.d = A.shape[1]
        self.M = A.shape[0]

        # Set up parameters
        self.Y = cp.Parameter(shape=(self.N, self.M), name="Y")
        self.A = cp.Parameter(shape=(self.M, self.d), name="A")
        
        self.Y.value = Y
        self.A.value = A
        
        # Set up variable
        self.R = cp.Variable(shape=(self.d,self.d),
                             nonneg=self.nonneg,
                             nonpos=self.nonpos,
                             symmetric=self.symmetric,
                             diag=self.diag,
                             PSD=self.PSD,
                             NSD=self.NSD,
                             boolean=self.boolean,
                             integer=self.integer,
                             pos=self.pos,
                             neg=self.neg)
        
        self.problem_is_set_up = True

    def solve(self,E,Y=None,A=None,penalty=0):
        
        if not self.problem_is_set_up:
            self.setup(Y=Y,A=A)
            
        # Set up losses 
        self.main_loss = cp.norm(x=self.Y - E @ self.R @ self.A.T, p=self.NORM)**2
        self.regularization_loss = cp.multiply(penalty, cp.norm(self.R, p=self.NORM)**2)

        # Combine losses        
        if penalty==0:
            self.L = self.main_loss
        else:
            self.L = self.main_loss + self.regularization_loss
        
        # Set up objective function
        self.Q = cp.Minimize(self.L)

        # Set up problem
        self.problem = cp.Problem(objective=self.Q)

        # Choose numerical solver
        self.solver = cp.SCS

        self.problem.solve(verbose=self.verbose, solver=self.solver, ignore_dpp=True)
        
        self.R_optim = self.R.value
        
        if sp.sparse.issparse(self.R_optim):
            self.R_optim = self.R_optim.toarray()
            
        return self.R_optim


***** Load data and solve problem *****
A = np.array(pd.read_parquet("https://www.dropbox.com/s/q6qlzwvgticoj8w/A.parquet?dl=1"))
E = np.array(pd.read_parquet("https://www.dropbox.com/s/k9oagw3ucezh7vz/E.parquet?dl=1"))
Y = np.array(pd.read_parquet("https://www.dropbox.com/s/pw45bd4pqc94n56/Y.parquet?dl=1"))

finetuner = FineTuner()    
R_opt = finetuner.solve(E=E, Y=Y, A=A)
```

**Comment by rileyjmurray** [comment]
Hi @muhlbach! Thanks for providing such a great code snippet to understand your problem. My first comment is that you shouldn't use cvxpy Parameter objects for this problem. It's far too large and cvxcore is almost certain to crash. What are the specs of the machine you're running this on? After replacing all cvxpy Parameters with plain old constants, cvxcore still consumed a ton of memory and crashed on my laptop.

Regarding specialized solvers, I think you should be able to reformulate the problem as the non-regularized version by working with augmented matrices for problem data. Consider the simpler problem of ridge regression:
```
min || A x - b||_2^2 + lamb * || x ||_2^2.
```
This can be rewritten as
```
min ||  [[A], [sqrt(lamb) * I]] x - [[b], [0]] ||_2^2 
```
where `` [[A], [sqrt(lamb) * I]]`` is a block matrix formed by stacking A on top of a scaled identity matrix, and ``[[b], [0]]`` is a block vector formed by stacking b on top of the zero vector.

But I also have a comment on your larger strategy of using regularization here. While it was me who suggested you use the Frobenius norm, your intuition of using the operator norm will likely lead to solutions that generalize far better to unseen data. I suggest that you test this hypothesis by running the model using the operator norm on a smaller dataset. Then, if you indeed get much better generalization, we can consider the (likely difficult) question of how to solve that problem at large scales.

**Comment by muhlbach** [comment]
@rileyjmurray Thank you so much for taking the time to look at my problem in this level of details--highly appreciated!

Let me address each of your comments one-by-one.

**Re: Parameter objects**
I've changed the Parameter objects to plain old constants, thanks! I originally adopted the Parameter objects in the case I needed to re-run the solver to tune a penalty hyperparameter by cross-validation but that seems a little infeasible at the moment.

**Re: Specs**
I'm running on a MacBook Pro, M1 (2020) with 8 cores and 16 GB ram. The above is running within acceptable time, but it does frighten me to implement some kind of cross-validation.

**Re: Reformulation**
This is a very good idea! I don't have too much experience in reformulating problems like this one, but that's definitely something to think about. The thing is that I would need to tune `lamb` by cross-validation, but if the problem gets formulated in an efficient way, it might not be a problem. Another problem with this strategy is that I do not have that much data, so splitting the data while still having to solve for `d` parameters (in the case I enforce `R` to be diagonal) is problematic (related to my last point, see bottom).

**Re: Choice of norm**
Very good point to try out different solvers. I'm going through the most obvious choices right now to test performance. It already appears that the nuclear norm is boosting performance out-of-sample, but still not enough to beat the identity matrix. This is of course not too scientific either because that would require another validation set, but that is something I can live with for now.

**Re: A follow-up question on the smaller problem vs full-size**
I'm currently testing different things on a smaller scale (with `N=400` compared to `900` in the full problem). I'm actually wondering whether the performance deteriorates/improves with the scale of `N`? For instance, if I estimate `R` using, say 400 observations, can I expect performance to be the same when using all observations such that the smaller problem is a fair proxy or is this generally not known? Makes little sense to test things on smaller scale if I generally can't expect these tests to be representative. Hope it makes sense!
